### ISSUE TITLE
Error: `Index X is out of bound for tuple` after a if-guard that guarantees in-bounds

### ISSUE THREAD

**Issue by jeertmans** [astral-sh/ty#1186]
### Summary

Hi!

Apparently, ty is unable to infer the length of tuple inside an `if`-guarded check that exactly checks the length:

```python
def f(tup: tuple[int, int] | tuple[int, int, int]) -> int:
    if len(tup) == 3:
        return tup[0] + tup[1] + tup[2]
    return tup[0] + tup[1]
```

Using `ty check` on this file raises:

```
error[index-out-of-bounds]: Index 2 is out of bounds for tuple `tuple[int, int]` with length 2
 --> issue.py:3:34
  |
1 | def f(tup: tuple[int, int] | tuple[int, int, int]) -> int:
2 |     if len(tup) == 3:
3 |         return tup[0] + tup[1] + tup[2]
  |                                  ^^^
4 |     return tup[0] + tup[1]
  |
info: rule `index-out-of-bounds` is enabled by default
```

I think ty should be able to infer the correct type of `tup` inside the `if` statement.

I searched the issue tracker for a similar bug report, but couldn't find any, so I hope this is not a duplicate :-)

Thanks in advance!

### Version

ty 0.0.1-alpha.20

**Comment by sharkdp** [comment]
> I searched the issue tracker for a similar bug report, but couldn't find any, so I hope this is not a duplicate :-)

I think https://github.com/astral-sh/ty/issues/560 asks for the same thing. But no worries, finding the right tickets with GitHub's search is challenging.

**Comment by jeertmans** [comment]
Thanks for pointing, and sorry for the duplicate!
